packages/babel-types/src/validators/isReferenced.ts

```diff
@@ -44,31 +44,34 @@ export default function isReferenced(
     case "ClassMethod":
     case "ClassPrivateMethod":
     case "ObjectMethod":
-      // @ts-expect-error todo(flow->ts) params have more specific type comparing to node
-      if (parent.params.includes(node)) {
-        return false;
+      if (parent.key === node) {
+        return !!parent.computed;
       }
-    // fall through
+      return false;
 
     // yes: { [NODE]: "" }
     // no: { NODE: "" }
     // depends: { NODE }
     // depends: { key: NODE }
-    // fall through
     case "ObjectProperty":
+      if (parent.key === node) {
+        return !!parent.computed;
+      }
+      if (parent.value === node) {
+        return !grandparent || grandparent.type !== "ObjectPattern";
+      }
+      return true;
```

Comment:
When can the node be neither the key nor the value?

Reply:
Good question! `node` could be neither `key` nor the `value` when `parent` is not the actual parent of `node`. This is actually ill-defined behaviour of `t.isReferenced(node, parent)`: It returns `true` if the `parent.type` is not handled by current logic and return `false` if handled since the conditions may never holds. Unless we change the interface to `isReferenced(NodePath)`, I don't think we can mitigate such issues.

I have updated `ObjectProperty` handling to align with other types such as `ClassProperty`.